jujusvg.go

```diff
@@ -13,8 +13,38 @@ import (
 // NewFromBundle returns a new Canvas that can be used
 // to generate a graphical representation of the given bundle
 // data. The iconURL function is used to generate a URL
-// that refers to an SVG for the supplied charm URL.
-func NewFromBundle(b *charm.BundleData, iconURL func(*charm.Reference) string) (*Canvas, error) {
+// that refers to an SVG for the supplied charm URL.  If
+// requested, icon SVGs that are available will be embedded
+// within the bundle SVG so that the bundle will be self-
+// contained as much as possible.  If that is to be the case,
+// utilize the provided IconFetcher (or default to HttpFetcher
+// if nil) to retrieve the icons.
+func NewFromBundle(b *charm.BundleData, iconURL func(*charm.Reference) string, embedIcons bool, fetcher IconFetcher) (*Canvas, error) {
```

Comment:
The interplay between iconURL, embedIcons and fetcher is a bit subtle here.

I still think that things could be simpler (and easier to explain) if
we separated the concerns of canvas creation from those
of embedding.

I'm thinking of something like this:

	// NewFromBundle returns a new Canvas that can be used
	// to generate a graphical representation of the given bundle
	// data. The iconURL function is used to generate a URL
	// that refers to an SVG for the supplied charm URL.
	func NewFromBundle(b *charm.BundleData, iconURL func(*charm.Reference) string) (*Canvas, error) {
		var canvas Canvas
	
		// Verify the bundle to make sure that all the invariants
		// that we depend on below actually hold true.
		if err := b.Verify(nil); err != nil {
			return nil, errgo.Notef(err, "cannot verify bundle")
		}
		// Go through all services in alphabetical order so that
		// we get consistent results.
		serviceNames := make([]string, 0, len(b.Services))
		for name := range b.Services {
			serviceNames = append(serviceNames, name)
		}
		sort.Strings(serviceNames)
		services := make(map[string]*service)
		for _, name := range serviceNames {
			serviceData := b.Services[name]
			x, xerr := strconv.ParseFloat(serviceData.Annotations["gui-x"], 64)
			y, yerr := strconv.ParseFloat(serviceData.Annotations["gui-y"], 64)
			if xerr != nil || yerr != nil {
				return nil, errgo.Newf("service %q does not have a valid position", name)
			}
			charmId, err := charm.ParseReference(serviceData.Charm)
			if err != nil {
				// cannot actually happen, as we've verified it.
				return nil, errgo.Notef(err, "cannot parse charm %q", serviceData.Charm)
			}
			svc := &service{
				name:      name,
				charmPath: charmId.Path(),
				point:     image.Point{int(x), int(y)},
				iconUrl:   iconURL(charmId),
			}
			services[name] = svc
			canvas.addService(svc)
		}
		for _, relation := range b.Relations {
			canvas.addRelation(&serviceRelation{
				serviceA: services[strings.Split(relation[0], ":")[0]],
				serviceB: services[strings.Split(relation[1], ":")[0]],
			})
		}
		return &canvas, nil
	}
	
	// FetchIcons fetches all the content for the charm icons
	// using the given fetcher and causes it to be embedded
	// directly into the generated SVG.
	func (c *Canvas) FetchIcons(fetcher IconFetcher) error {
		urlMap := make(map[string]int)
		urls := make([]string, 0, len(c.services))
		for _, svc := range c.services {
			if _, ok := urlMap[svc.iconUrl]; ok {
				continue
			}
			urlMap[svc.iconUrl] = len(urls)
			urls = append(urls, svc.iconUrl)
		}
		content, err := fetcher.FetchIcons(urls)
		if err != nil {
			return err
		}
		for _, svc := range c.services {
			svc.iconSrc = content[urlMap[svc.iconUrl]]
		}
		return nil
	}

which allows the IconFetcher interface to become somewhat simpler:

	// An IconFetcher provides functionality for retrieving icons for the charms
	// within a given bundle.  The FetchIcons method accepts a slice of icon
	// URLs and returns a corresponding slice of icon contents, one element
	// for each URL.
	type IconFetcher interface {
		FetchIcons(urls []string) ([][]byte, error)
	}

which allows the implementation of the fetcher to become simpler too
(no need to do deduplication within the fetcher itself)

What do you think?